src/empirical.jl

```diff
@@ -1,68 +1,142 @@
+    end
 end
 
-function (ecdf::ECDF)(v::RealVector)
-    evenweights = isempty(ecdf.weights)
-    weightsum = evenweights ? length(ecdf.sorted_values) : sum(ecdf.weights)
-    ord = sortperm(v)
-    m = length(v)
-    r = similar(ecdf.sorted_values, m)
-    r0 = zero(weightsum)
-    i = 1
-    for (j, x) in enumerate(ecdf.sorted_values)
-        while i <= m && x > v[ord[i]]
-            r[ord[i]] = r0
-            i += 1
-        end
-        r0 += evenweights ? 1 : ecdf.weights[j]
-        if i > m
-            break
-        end
-    end
-    while i <= m
-        r[ord[i]] = weightsum
-        i += 1
+# broadcasts ecdf() over an array
+# caches the last calculated value
+function Base.Broadcast.broadcasted(ecdf::ECDF, v::AbstractArray)
+    res = similar(v, weighttype(ecdf))
+    @inbounds for i in eachindex(v)
+        res[i] = i == 1 || v[i] != v[i-1] ? ecdf(v[i]) : res[i-1]
     end
-    return r / weightsum
+    return res
 end
 
 """
-    ecdf(X; weights::AbstractWeights)
+    ecdf(X::RealVector; weights::AbstractWeights=nothing, interpolate::Bool=false) -> ECDF
+
+Construct an *empirical cumulative distribution function* (ECDF) based on a vector of samples (`X`).
+Returns a callable [`ECDF`](@ref) object.
 
-Return an empirical cumulative distribution function (ECDF) based on a vector of samples
-given in `X`. Optionally providing `weights` returns a weighted ECDF.
+# Example
 
-Note: this function that returns a callable composite type, which can then be applied to
-evaluate CDF values on other samples.
+```jldoctest
+julia> using StatsBase
 
-`extrema`, `minimum`, and `maximum` are supported to for obtaining the range over which
-function is inside the interval ``(0,1)``; the function is defined for the whole real line.
+julia> X = 1:15
+1:15
+
+julia> xcdf = ecdf(X)
+ECDF{Int64,Float64,false}(15 values)
+
+julia> xcdf(9)
+0.6
+```
+
+## Keyword arguments
+* `weights`: optional weights for each `X` sample
+* `interpolate`: when enabled, `ECDF(y)` will linearly interpolate the cumulative density between
+  the neighboring `X` samples (``x_i \\leq y < x_{i+1}``),
+  otherwise (by default) `ECDF(x[i])` is returned
+
+## Notes
+The returned function is defined for the whole real line.
+Use [`extrema`](@ref), [`minimum`](@ref), and [`maximum`](@ref) to obtain the range
+over which the ECDF is inside the interval ``(0, 1)``.
 """
-function ecdf(X::RealVector; weights::AbstractVector{<:Real}=Weights(Float64[]))
+function ecdf(X::RealVector;
+              weights::Union{Nothing, RealVector}=nothing,
+              interpolate::Bool=false)
     any(isnan, X) && throw(ArgumentError("ecdf can not include NaN values"))
-    isempty(weights) || length(X) == length(weights) || throw(ArgumentError("data and weight vectors must be the same size," *
-        "got $(length(X)) and $(length(weights))"))
+    evenweights = isnothing(weights) || isempty(weights)
```

Comment:
Inference works better with `===`:
```suggestion
    evenweights = weights === nothing || isempty(weights)
```